### ISSUE TITLE
Types which differ only in `_` suffix can lead to e.g., duplicate variables in code generation

### ISSUE THREAD

**Issue by bbannier** [zeek/spicy#1499]
If a module contains two types with names which only differ in a `_` suffix (e.g., `D` and `D_`) can lead to broken code.

@Mohan-Dhawan provided the following example:

```ruby
# foo.spicy
module foo;

import bar;

public type C = unit {
    x: bar::D_;
};
```

```ruby
# bar.spicy
module bar;

public type D = unit {};
public type D_ = unit {};
```

```console
# spicyc -dj foo.spicy bar.spicy -o baz.hlto
[internal-error] <import>: constant '__hlt::type_info::::__ti_bar_D' already exists differently in C++ translation unit

--- Aborting
# 0   libhilti.dylib                      0x000000010df1c9db _ZN5hilti2rt9BacktraceC2Ev + 235
# 1   libhilti.dylib                      0x000000010dac817b _ZN5hilti4util20abort_with_backtraceEv + 139
# 2   libhilti.dylib                      0x000000010dac2d98 _ZN5hilti6Logger13internalErrorERKNSt3__112basic_stringIcNS
1_11char_traitsIcEENS1_9allocatorIcEEEERKNS_8LocationE + 82
# 3   libhilti.dylib                      0x000000010dbc424a _ZN5hilti6detail3cxx4Unit3addERKNS1_11declaration8ConstantE
RKNS_4MetaE + 342
# 4   libhilti.dylib                      0x000000010dbc87e1 _ZN5hilti6detail3cxx4Unit18importDeclarationsERKS2_ + 329
# 5   libhilti.dylib                      0x000000010dc1f288 _ZN5hilti4Unit7codegenEv + 2134
# 6   libhilti.dylib                      0x000000010dbdfcc4 _ZN5hilti6Driver13_codegenUnitsEv + 720
# 7   libhilti.dylib                      0x000000010dbe0be7 _ZN5hilti6Driver12compileUnitsEv + 1547
# 8   libhilti.dylib                      0x000000010dbe120d _ZN5hilti6Driver7compileEv + 31
# 9   libhilti.dylib                      0x000000010dbe0f01 _ZN5hilti6Driver3runEv + 117
# 10  spicyc                              0x000000010c1cb6df main + 271
# 11  dyld                                0x00007ff804a9341f start + 1903
```

A workaround is to ensure that a modules contains only types that have distinct names, even when stripping a `_` suffix.

**Comment by bbannier** [comment]
A likely seems to be `util::toIdentifier` and how it uses `_` for normalizing characters. C++ also allows some unicode in identifiers (with potentially spotty compiler support though depending on the compiler version), so we might be able to use a character inaccessible to users instead.

**Comment by bbannier** [comment]
#1504 added the option to use `@` in identifiers and have them properly normalized.

**Comment by bbannier** [comment]
It looks like this will be fixed by #1462.

**Comment by rsmmr** [comment]
Indeed, this is fixed now.
